### PR TITLE
WIP: Prototype for monitoring of cartographer.

### PR DESCRIPTION
**Pull request by stevewolter**
This PR is a prototype for illustrating an RFC, not a request for submission. Please keep code reviews to a high level.

This PR introduces an interface for recording histograms with pre-defined buckets in the cartographer core library, and a factory for producing instances of these histograms. It also adds an implementation in cartographer_grpc that delegates to the prometheus client library.

### CODE REVIEW

File: cartographer_grpc/metrics/prometheus/family_factory.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#include <string>
+
+#include "cartographer/metrics/family_factory.h"
+#include "prometheus/registry.h"
+
+namespace cartographer_grpc {
+namespace metrics {
+namespace prometheus {
+
+class FamilyFactory : public cartographer::metrics::FamilyFactory {
```
Comment (gaschler) [inline cartographer_grpc/metrics/prometheus/family_factory.h]: naming: I know this is already in the namespace, but I find MetricFamilyFactory more informative.

File: cartographer/metrics/family_factory.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#include "cartographer/metrics/histogram.h"
+
+namespace cartographer {
+namespace metrics {
+
+class FamilyFactory {
+ public:
+  virtual ~FamilyFactory() = default;
+
+  virtual HistogramFamily* NewHistogramFamily(
```
Comment (gaschler) [inline cartographer/metrics/family_factory.h]: Looks good, this interface is not intrusive at all.

**Comment by SirVer** [comment]
What is prometheus?

File: cartographer_grpc/metrics/prometheus/family_factory.cc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+class HistogramFamily : public cartographer::metrics::HistogramFamily {
+ public:
+  HistogramFamily(::prometheus::Family<::prometheus::Histogram>* prometheus)
+      : prometheus_(prometheus) {}
+
+  Histogram* Add(const std::map<std::string, std::string>& labels,
+                 const Histogram::BucketBoundaries& boundaries) override {
+    ::prometheus::Histogram* histogram = &prometheus_->Add(labels, boundaries);
+    auto wrapper = new Histogram(histogram);
```
Comment (MichaelGrupp) [inline cartographer_grpc/metrics/prometheus/family_factory.cc]: Any reason for a naked `new` call?

**Comment by slamer-run** [comment]
thank you for implementing to monitor metrics. 
Referring to "Implementing the facade" at below link, the metrics information can be seen through "http://localhost:9100/metrics" with grpc_demo_backpack_2d.launch.

https://github.com/googlecartographer/rfcs/blob/master/text/0014-monitoring.md

However, I can't check with diagram from a nice dashboard. Should I use "prometheus-to-sd" to see the figures about pose graph metrics and loop close counter?  Could you tell me how to use "prometheus-to-sd" ?

**Comment by cschuet** [comment]
@slamer-run What you see under "http://localhost:9100/metrics" is "raw" [Prometheus](https://prometheus.io/) data, i.e. you are now in the Prometheus ecosystem and there are lots of tools to generate the visualizations.

One option is to use Stackdriver for visualization: prometheus-to-sd is a scraper that scrapes the information from "http://localhost:9100/metrics" on a regular basis and exports them to Stackdriver (which gives you nice graphs). Note however that out-of-the-box prometheus-to-sd only runs on the GCE/GKE and you some hackery to make it run on your robot (see [here](https://github.com/stevewolter/k8s-stackdriver)).

Alternatively there are lots of open source tools for scraping Prometheus data and building cool visualizations for it, see e.g. [here](https://prometheus.io/docs/visualization/grafana/).

**Comment by slamer-run** [comment]
@cschuet 
Thank you for your suggestion!
Though I tried grafana as you suggested me, I haven't seen the result at nice dashboard. Probably, my configuration of grafana is not correct... Since I haven't been familiar with infrastructure engineering tools, I'm going to see study prometeus and grafana in this weekend!

PS: If you know good manual to check metrics result with grafana, let me know it.

**Pull request closed without merging**